### PR TITLE
Issue #46 feat: Enhance pipeline jobs to push all failed and invalid events into single topic

### PR DESCRIPTION
**Pull request by shiva-rakshith**
**Expected outcome:**
Enhance all the pipeline jobs to write failed, invalid and duplicate events to a single topic. So that data can be queried from a single dataset to analyse the failed, invalid and duplicate events.

The following are the changes made as part of this PR:

1. Updated the failed event structure.
2. Write failed events from extractor, pre-processor, denormalizer, transformer and druid-router to a single failed topic. 
3. Write all master data processor failed events to master data failed topic.
4. Added error code and exception for master data config key missing in masterdata processor job.
5. Added exception handling, if an invalid JSON is ingested to ingest kafka topic.

**Test Scenarios:** 
The following test scenarios are applicable for both dataset and master dataset events.

- [x]  Scenario 1: When an event without extraction key is ingested, it should go to the failed topic.
- [x]  Scenario 2: When a duplicate event is ingested, it should go to the failed topic.
- [x]  Scenario 3: When an event with invalid schema is ingested, it should go to the failed topic.
- [x]  Scenario 4: When an event is ingested and denormalisation is failed, it should go to the failed topic.
- [x]  Scenario 5: When a batch event is ingested without dataset id or dataset configuration or extraction configuration, the failed event contains the original event, as there won't be an extraction key to remove the original event from the failed.
- [x] Scenario 6: When an invalid JSON is ingested directly to Kafka, pipeline should through invalid JSON exception and write to failed topic.
- [x] Scenario 7: When a master dataset event is ingested without master dataset configuration, it should through an exception and write to failed topic.

### CODE REVIEW

**Review by ManjunathDavanam — CHANGES_REQUESTED** [review]
1. Please create an ingestion spec for the failed events submission
2. Please automate the ingestion spec submission through the helm chart - Ref (https://github.com/Sanketika-Obsrv/obsrv-automation/blob/main/terraform/modules/helm/submit_ingestion/submit-ingestion-helm-chart/templates/config-map.yaml)
3. Please automate the dashboard creation for the failed events summary.  (Ref. https://github.com/Sanketika-Obsrv/obsrv-automation/pull/19/files)

**Review by manjudr — COMMENTED** [review]
1. Please create an ingestion spec for the failed events submission
2. Please automate the ingestion spec submission through the helm chart - Ref (https://github.com/Sanketika-Obsrv/obsrv-automation/blob/main/terraform/modules/helm/submit_ingestion/submit-ingestion-helm-chart/templates/config-map.yaml)
3. Please automate the dashboard creation for the failed events summary.  (Ref. https://github.com/Sanketika-Obsrv/obsrv-automation/pull/19/files)

**Comment by manjudr** [comment]
Please add the description and standardise all the commits going forward.

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala [hunk 1]
```diff
@@ -147,7 +161,11 @@ class ExtractionFunction(config: ExtractorConfig, @transient var dedupEngine: De
     wrapperEvent
   }
 
-  private def markBatchFailed(batchEvent: mutable.Map[String, AnyRef], error: Error): mutable.Map[String, AnyRef] = {
+  private def markBatchFailed(batchEvent: mutable.Map[String, AnyRef], error: Error, extractionKey: String): mutable.Map[String, AnyRef] = {
+    if (!extractionKey.isEmpty) {
```
Comment (manjudr) [inline pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:165]:
```
copiedData = batchEvent.get(extractionKey)
batchEvent.remove(extractionKey)
batchEvent.put(Constants.EVENT, copiedData)
```

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala [hunk 2]
```diff
@@ -37,26 +40,34 @@ class ExtractionFunction(config: ExtractorConfig, @transient var dedupEngine: De
                               context: ProcessFunction[mutable.Map[String, AnyRef], mutable.Map[String, AnyRef]]#Context,
                               metrics: Metrics): Unit = {
     metrics.incCounter(config.defaultDatasetID, config.totalEventCount)
+    if(batchEvent.contains(Constants.INVALID_EVENT)) {
```
Comment (manjudr) [inline pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:43]: Handle with the error code from the obsrv_meta

**Review by manjudr — APPROVED** [review]

File: framework/src/main/scala/org/sunbird/obsrv/core/model/Constants.scala [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package org.sunbird.obsrv.core.model
+
+object Constants {
+
+  val INVALID_EVENT = "invalidEvent"
```
Comment (manjudr) [inline framework/src/main/scala/org/sunbird/obsrv/core/model/Constants.scala:5]: Please remove this

**Review by manjudr — CHANGES_REQUESTED** [review]

File: framework/src/main/scala/org/sunbird/obsrv/core/model/Constants.scala (hunk 3 above)
Comment (shiva-rakshith) [reply framework/src/main/scala/org/sunbird/obsrv/core/model/Constants.scala]: removed

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala (hunk 2 above)
Comment (shiva-rakshith) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala]: updated the code

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala (hunk 1 above)
Comment (shiva-rakshith) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala]: updated the code

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala [hunk 4]
```diff
@@ -37,26 +41,35 @@ class ExtractionFunction(config: ExtractorConfig, @transient var dedupEngine: De
                               context: ProcessFunction[mutable.Map[String, AnyRef], mutable.Map[String, AnyRef]]#Context,
                               metrics: Metrics): Unit = {
     metrics.incCounter(config.defaultDatasetID, config.totalEventCount)
+    val obsrvMeta = batchEvent.clone()(Constants.OBSRV_META).asInstanceOf[mutable.Map[String, AnyRef]]
```
Comment (manjudr) [inline pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:44]: Can we avoid this clone object?

Comment (manjudr) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:44]: Clone of object could be expensive operation.

Comment (manjudr) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:44]: Or clone the object only when it's required.

Comment (shiva-rakshith) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala]: Removed cloning the object

File: pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala [hunk 5]
```diff
@@ -147,8 +156,10 @@ class ExtractionFunction(config: ExtractorConfig, @transient var dedupEngine: De
     wrapperEvent
   }
 
-  private def markBatchFailed(batchEvent: mutable.Map[String, AnyRef], error: Error): mutable.Map[String, AnyRef] = {
+  private def markBatchFailed(batchEvent: mutable.Map[String, AnyRef], error: Error, extractionKey: String): mutable.Map[String, AnyRef] = {
     super.markFailed(batchEvent, error, config.jobName)
+    if (extractionKey.nonEmpty && StringUtils.equals(extractionKey, Constants.EVENT))
+      batchEvent.remove(extractionKey)
```
Comment (sowmya-dixit) [inline pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:162]: Why is extractionKey removed from batchEvent?

Comment (shiva-rakshith) [reply pipeline/extractor/src/main/scala/org/sunbird/obsrv/extractor/functions/ExtractionFunction.scala:162]: We are remove the event from root level and adding as stringified event in the failed.

Following is the structure of failed event:
```
{
    "dataset-id": "dataset-id",
    "obsrv_meta": {
        "flags": { "ExtractorJob": "failed" },
        "syncts": 1696504444806,
        "prevProcessingTime": 1696504445231,
        "error": {
            "src": "ExtractorJob",
            "error_code": "ERR_EXT_1001",
            "error_msg": "Unable to extract the data from the extraction key"
        },
        "processingStartTime": 1696504445048,
        "timespans": { "ExtractorJob": 183 },
        "source": {
            "meta": {
                "id": "<dataset_source_config_id>", # For API this wouldn't be present
                "connector_type": "kafka/api/jdbc",
                "version": "1.0",
                "entry_source": <Table/Topic/API>
            },
            "trace_id": "<uuid>" # <uniquely_trace_event>
        }
    },
    "event": <STRINGIFIED JSON OF THE ENTIRE RECEIVED MESSAGE>
}
```

**Pull request merged**